src/board/system76/common/flash/main.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2020 Evan Lojewski
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <stdint.h>
+#include <board/flash.h>
+
+// EC indirect flash access
+volatile uint8_t __xdata __at(0x103B) ECINDAR0;
+volatile uint8_t __xdata __at(0x103C) ECINDAR1;
+volatile uint8_t __xdata __at(0x103D) ECINDAR2;
+volatile uint8_t __xdata __at(0x103E) ECINDAR3;
+volatile uint8_t __xdata __at(0x103F) ECINDDR;
+
+#define SPI_DEVICE                  (0x70)
+#define SPI_FOLLOW_MODE             (0x0F)
+#define SPI_CHIP_SELECT             (0xFD)
+#define SPI_CHIP_DESELECT           (0xFE)
+
+#define SPI_READ_STATUS_COMMAND     (0x05)
+#define SPI_READ_COMMAND            (0x0B)
+#define SPI_WRITE_COMMAND           (0x02)
+#define SPI_WRITE_ENABLE_COMMAND    (0x06)
+#define SPI_READ_STATUS_COMMAND     (0x05)
+
+#define SPI_ERASE_SECTOR_COMMAND    (0xD7)
+
+#define SPI_STATUS_WIP              (0x01)
+
+void flash_enter_follow_mode(void);
+void flash_exit_follow_mode(void);
+void flash_wait(void);
+void flash_write_enable(void);
+
+/**
+ * Main flash API entry point.
+ *
+ * NOTE: This *must* be the first function in this file to ensure that it is placed
+ *          first in the resulting binary. This is required to ensure that address
+ *          matches the address (FLASH_OFFSET) for flash_entry in wrapper.c.
+ * NOTE: __reentrant so that parameters and temperary vairables are placed on the
+ *          stack, ensuring the main application __data variables are not stomped on.
+ * NOTE: __critical to ensure interrupts are disabled. This does mean that interrupt
+ *          such as the timer will be block until flash acccess is complete
+ */
+void flash_entry(uint32_t addr, uint8_t * data, uint8_t length, uint8_t command) __reentrant __critical {
```

Comment:
If we don't expect to use this for flashing the code space below 64 KiB, perhaps doing this would be good:

```
if (addr < 0x10000 || addr >= 0x20000) return;
```

Reply:
Adding a check like this sounds good to me. If we ever want it to write to the lower 64KB, the code can always change at that point.
I'll also look at making sure the length doesn't cause it to overflow.

Reply:
I've added this check with an additional check for addr + length.
Note that both addr and addr + length are checked to ensure there's no wrapping. (MAX_UINT start address wrapping to 0)